primitives/src/merkle_tree/field_based_mht/smt/big_merkle_tree.rs

```diff
@@ -0,0 +1,1044 @@
+    state_path: Option<String>,
+    // tree in-memory state
+    state: BigMerkleTreeState<F, T>,
+    // the height of the Merkle tree
+    height: usize,
+    // path to the db
+    path_db: String,
+    // stores the leaves
+    database: DB,
+    // path to the cache
+    path_cache: String,
+    // stores the cached nodes
+    db_cache: DB,
+
+    _field: PhantomData<F>,
+    _parameters: PhantomData<T>,
+    _poseidon_parameters: PhantomData<P>,
+}
+
+impl<F: PrimeField + MulShort, T: SmtPoseidonParameters<Fr=F>, P: PoseidonParameters<Fr=F>> Drop for BigMerkleTree<F, T, P> {
+    fn drop(&mut self) {
+        self.close();
+    }
+}
+
+impl<F: PrimeField + MulShort, T: SmtPoseidonParameters<Fr=F>, P: PoseidonParameters<Fr=F>> BigMerkleTree<F, T, P> {
+    // Creates a new tree of specified `width`.
+    // If `persistent` is specified, then DBs will be kept on disk and the tree state will be saved
+    // so that the tree can be restored any moment later. Otherwise, no state will be saved on file
+    // and the DBs will be deleted.
+    pub fn new_unitialized(width: usize, persistent: bool, state_path: Option<String>, path_db: String, path_cache: String) -> Result<Self, Error> {
+
+        // If the tree must be persistent, that a path to which save the tree state must be
+        // specified.
+        if !persistent { assert!(state_path.is_none()) } else { assert!(state_path.is_some()) }
+
+        let height = width as f64;
+        let height = height.log(T::MERKLE_ARITY as f64) as usize;
+        let state = BigMerkleTreeState::<F, T>::get_default_state(width, height);
+        let path_db = path_db;
+        let database = DB::open_default(path_db.clone())
+            .map_err(|e| Error::Other(e.to_string()))?;
+        let path_cache = path_cache;
+        let db_cache = DB::open_default(path_cache.clone())
+            .map_err(|e| Error::Other(e.to_string()))?;
+        Ok(BigMerkleTree {
+            persistent,
+            state_path,
+            state,
+            height,
+            path_db,
+            database,
+            path_cache,
+            db_cache,
+            _field: PhantomData,
+            _parameters: PhantomData,
+            _poseidon_parameters: PhantomData,
+        })
+    }
+
+    // Creates a new tree starting from state at `state_path` and DBs at `path_db` and `path_cache`.
+    // The new tree may be persistent or not, the actions taken in both cases are the same as
+    // in `new_unitialized()`.
+    pub fn new(persistent: bool, state_path: String, path_db: String, path_cache: String) -> Result<Self, Error> {
+
+        let state = {
+            let state_file = fs::File::open(state_path.clone())
+                .map_err(|e| Error::Other(e.to_string()))?;
+            BigMerkleTreeState::<F, T>::read(state_file)
+        }.map_err(|e| Error::Other(e.to_string()))?;
+
+        //TODO: This is the "depth" more than the "height". Which should we keep ?
```

Comment:
This is the depth more than the height. Probably we should keep the notion of height, i.e. the number of levels from leaves to the root, and the number of leaves will be 2^(HEIGHT-1)